src/data/mv_polynomial/variables.lean

```diff
@@ -103,7 +103,7 @@ lemma degrees_X' (n : σ) : degrees (X n : mv_polynomial σ R) ≤ {n} :=
 le_trans (degrees_monomial _ _) $ le_of_eq $ to_multiset_single _ _
 
 @[simp] lemma degrees_X [nontrivial R] (n : σ) : degrees (X n : mv_polynomial σ R) = {n} :=
-(degrees_monomial_eq _ _ one_ne_zero).trans (to_multiset_single _ _)
+(degrees_monomial_eq _ (1 : R) one_ne_zero).trans (to_multiset_single _ _)
```

Comment:
Similarly, the elaborator needs a bit of help here.

Reply:
From `set_option trace.type_context.is_def_eq_detail true`, it looks like the following is happening:

Previously, `one_ne_zero` had the type  `_ ≠ (@has_zero.zero _ (@mul_zero_class.to_has_zero _ (monoid_with_zero.to_mul_zero_class _)))`, exactly what we're looking for. In this PR the goal is instead `_ ≠ @has_zero.zero _ (@mul_zero_class.to_has_zero _ (non_unital_non_assoc_semiring.to_mul_zero_class _)`. So the elaborator starts unfolding a lot of definitions. I think because it unfolds a lot of these definitions, it doesn't get around to unifying this type with the expected type from `trans _ (to_multiset_single _ _)`. So it picks the default instance for `mul_zero_class`, namely `nat`, and continues. Then it needs to check `trans _ (to_multiset_single _ _)`, sees `mv_polynomial` on one side and `nat` on the other, and errors.

Reply:
I think the only way to fix this (without touching the elaborator) is to make the `semiring` to `has_zero` diamond go via `monoid_with_zero`. But that is exactly what this PR is supposed to avoid so I think the conclusion is to insert the `(1 : R)` hint.

Reply:
The direction taken by the diamond can be adjusted by reordering the base classes + priorities I think, which doesn't go against the goal of the PR

Reply:
To be a bit more precise, I think the difference is that we need higher-order unification after this change, and before we only needed first-order unification. And because the higher-order unifier cannot be as powerful as the first-order unifier, it needs more hints.

Reply:
Also, that's not too say doing that reordering is necessarily worth the effort.

Reply:
Thanks @Vierkantor and @eric-wieser for these remarks.

Reply:
I'll go ahead and mark this conversation as resolved since we are OK with inserting the hint.